airbyte-integrations/connectors/source-zendesk-sunshine/source_zendesk_sunshine/streams.py

```diff
@@ -0,0 +1,210 @@
+from airbyte_cdk.sources.streams.http import HttpStream
+
+
+class SunshineStream(HttpStream, ABC):
+    primary_key = "id"
+    data_field = "data"
+    url_base = "https://XXX"
+    page_size = 100
+
+    def __init__(self, subdomain: str, start_date: pendulum.datetime, **kwargs):
+        self.url_base = f"https://{subdomain}.zendesk.com/api/sunshine/"  # url_base #kwargs.pop("base_url")
+        self._start_date = start_date
+        self.subdomain = subdomain
+        super().__init__(**kwargs)
+
+    def backoff_time(self, response: requests.Response) -> Optional[float]:
+        delay_time = response.headers.get("Retry-After")
+        if delay_time:
+            return int(delay_time)
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        resp_json = response.json()
+        if resp_json.get("links") and resp_json.get("links").get("next"):
+            next_query_string = urllib.parse.urlsplit(resp_json.get("links").get("next")).query
+            params = dict(urllib.parse.parse_qsl(next_query_string))
+            return params
+        return {}
+
+    def request_headers(self, **kwargs) -> Mapping[str, Any]:
+        return {"Content-Type": "application/json"}
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        yield from response_json.get(self.data_field, [])
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+
+        params = {"per_page": self.page_size}
+        if next_page_token:
+            params.update(next_page_token)
+        return params
+
+
+class IncrementalSunshineStream(SunshineStream, ABC):
+    state_checkpoint_interval = 1000
+    cursor_field = "updated_at"  # most common
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        """
+        Return the latest state by comparing the cursor value in the latest record with the stream's most recent state object
+        and returning an updated state object.
+        """
+        latest_state = latest_record.get(self.cursor_field)
+        current_state = current_stream_state.get(self.cursor_field) or latest_state
+        # dates are ISOformatted, no need to parse
+        return {self.cursor_field: max(latest_state, current_state)}
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        yield from response_json.get(self.data_field, [])
+
+
+class ObjectTypes(SunshineStream):
+    def path(self, **kwargs) -> str:
+        return "objects/types"
+
+
+class ObjectRecords(IncrementalSunshineStream):
+    http_method = "POST"
```

Comment:
Why POST is used?

Reply:
the get is valid only for full-refresh style. The is not possible to set date filter during get request. To make stream incremental, I must use another endpoint (`query` instead of `objects/records`) and set date filters there, because this is only the way that is documented. Since there is impossible to apply date filters I had to remove incremental from 2 or 3 other streams, which looked like support incremental but they are not.

Reply:
can you leave a comment explaining summarising this? it can also be a link to a spec. the goal is to make sure a reader can understand this as easily as possible.